File: card-words/src/main/java/com/thuanthichlaptrinh/card_words/core/domain/UserVocabProgress.java
package com.thuanthichlaptrinh.card_words.core.domain;

import com.thuanthichlaptrinh.card_words.common.enums.VocabStatus;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;
import java.util.UUID;

import org.hibernate.annotations.GenericGenerator;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "user_vocab_progress", indexes = {
        @Index(name = "idx_uvp_user_id", columnList = "user_id"),
        @Index(name = "idx_uvp_vocab_id", columnList = "vocab_id"),
        @Index(name = "idx_uvp_status", columnList = "status"),
        @Index(name = "idx_uvp_next_review_date", columnList = "next_review_date"),
        @Index(name = "idx_uvp_user_vocab", columnList = "user_id,vocab_id", unique = true)
})
public class UserVocabProgress extends BaseEntity {
    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @Column(name = "id", updatable = false, nullable = false, columnDefinition = "uuid")
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "vocab_id", nullable = false)
    private Vocab vocab;

    @Enumerated(EnumType.STRING)
    @Column(length = 50)
    private VocabStatus status;

    @Column(name = "last_reviewed")
    private LocalDate lastReviewed;

    @Builder.Default
    @Column(name = "times_correct", nullable = false)
    private Integer timesCorrect = 0;

    @Builder.Default
    @Column(name = "times_wrong", nullable = false)
    private Integer timesWrong = 0;

    @Builder.Default
    @Column(name = "ef_factor", nullable = false)
    private Double efFactor = 2.5;

    @Builder.Default
    @Column(name = "interval_days", nullable = false)
    private Integer intervalDays = 1;

    @Builder.Default
    @Column(nullable = false)
    private Integer repetition = 0;

    @Column(name = "next_review_date")
    private LocalDate nextReviewDate;
}

File: card-words-ai/app/db/models/user_vocab_progress.py
"""UserVocabProgress SQLAlchemy model"""
from sqlalchemy import Column, String, Integer, Float, Date, ForeignKey, Enum as SQLEnum
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
import uuid
import enum

from .base import Base


class VocabStatus(str, enum.Enum):
    """Vocabulary status enum"""
    NEW = "NEW"
    UNKNOWN = "UNKNOWN"
    KNOWN = "KNOWN"
    MASTERED = "MASTERED"


class UserVocabProgress(Base):
    """User vocabulary progress model"""
    __tablename__ = 'user_vocab_progress'
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    vocab_id = Column(UUID(as_uuid=True), ForeignKey('vocab.id'), nullable=False, index=True)
    
    status = Column(SQLEnum(VocabStatus), nullable=False, index=True)
    last_reviewed = Column(Date)
    times_correct = Column(Integer, default=0, nullable=False)
    times_wrong = Column(Integer, default=0, nullable=False)
    ef_factor = Column(Float, default=2.5, nullable=False)
    interval_days = Column(Integer, default=1, nullable=False)
    repetition = Column(Integer, default=0, nullable=False)
    next_review_date = Column(Date, index=True)
    
    # Relationship
    vocab = relationship("Vocab", lazy="joined")
    
    def __repr__(self):
        return f"<UserVocabProgress(user_id={self.user_id}, vocab_id={self.vocab_id}, status={self.status})>"
